### ISSUE TITLE
With --udp and --reverse, out-of-order packet reception corrupts sequence numbers / test

### ISSUE THREAD

**Issue by aahoward** [esnet/iperf#1123]
_NOTE: The iperf3 issue tracker is for registering bugs, enhancement
requests, or submissions of code.  It is not a means for asking
questions about building or using iperf3.  Those are best directed
towards the iperf3 mailing list at iperf-dev@googlegroups.com or
question sites such as Stack Overflow
(http://www.stackoverflow.com/).  A list of frequently-asked questions
regarding iperf3 can be found at http://software.es.net/iperf/faq.html._

# Context

* Version of iperf3: 3.9

* Hardware: X86

* Operating system (and distribution, if any): Linux Ubuntu [ multiple ]

# Bug Report
A WAN connection between our datacentre and  AWS exhibits high packet out of order characteristics. When using iperf3 UDP reverse test mode with client on prem and server in AWS we observe corruption of packet counters. The following can be used to reproduce the problem on demand within our network.

```
iperf39 -u -c aws_ec2_ip -l 100 -t 1 -b 10k -R

Note. this will only error when out of order packet delivery is common
```

A test script has been included to automate the reproduction of this bug, In addition a PCAP and JSON output is included that show an incident of the bug.

The bug is caused by frame1 [4 bytes server response to client ] and frame2 [ first test payload ] being recieved in reverse order.  

The client reads 4 bytes from the first frame recieved [ frame2 ] , leaving 96 bytes remaining , then read's 100 bytes [ 96+4 ]. This results in a framing error when the counters are memcpy from the buffer.

Introducing a sleep in iperf_udp.c after the 4 bytes response has been transmitted will allow sufficient time to ensure that the 4 byte response and the first test load are never received out of sequence. A 10ms sleep should be more than enough to cater for most network conditions with respect to out of order packets. 

This bug is only applicable to out of order in the context of the server 4 byte "ping/pong" response and the first testing payload.

* Expected Behavior
4 byte response "987654321" response should always arrive before first test payload packet
* Actual Behavior
Testing AWS to onpremise network links show high amounts of out of order frames. On multiple occasions the 4 byte response from the server to the client arrives after the first payload packet. This creates framing issues.
* Steps to Reproduce
Script/JSON results and PCAP have been attached that show the problem
* Possible Solution
Introduce a micro sleep statement after the 4 byte server response has been generated in  iperf_udp_accept iperf_udp.c .  Testing with 10ms sleep elliminated the problem for me , perhaps this should be configurable as a command line option.
.
[issue-files.tar.gz](https://github.com/esnet/iperf/files/5972050/issue-files.tar.gz)

**Comment by davidBar-On** [comment]
@bmah888, the analysis done by @aahoward seem to be very important, as it can also explain the root cause of other issues, such as #914.

Is the suggested workaround to add a delay after sending the 4 bytes is good enough?  Is a more robust solution required, e.g. sending the first UDP message instead of the 4 bytes?

**Comment by bmah888** [comment]
I'm thinking about this one, admittedly I am just reading through it for the first time. It seems like an unlikely case, but when it does happen, the effects are pretty severe. It's notable that this is another problem that happens with `--reverse`.

I'm not sure if a simple delay really eliminates the hazard, as we don't know when the delay would be "enough". Again I haven't looked through the supporting documentation yet, so the doubt that I express is admittedly not evidence-based.

**Comment by davidBar-On** [comment]
Based on @aahoward analysis in this issue, submitted a proposed fix in PR #1260.

**Comment by bmah888** [comment]
Believed fixed by #1260.
